src/cargo/core/workspace.rs

```diff
@@ -1522,27 +1519,83 @@ impl<'gctx> Workspace<'gctx> {
             })
             .sorted()
             .take(5)
-            .collect();
+            .collect()
+    }
 
+    fn report_unknown_features_error(
+        &self,
+        specs: &[PackageIdSpec],
+        cli_features: &CliFeatures,
+        found_features: &BTreeSet<FeatureValue>,
+    ) -> CargoResult<()> {
         let unknown: Vec<_> = cli_features
             .features
             .difference(found_features)
             .map(|feature| feature.to_string())
             .sorted()
             .collect();
 
-        if suggestions.is_empty() {
-            bail!(
-                "none of the selected packages contains these features: {}",
+        let (selected_members, unselected_members): (Vec<_>, Vec<_>) = self
+            .members()
+            .partition(|member| specs.iter().any(|spec| spec.matches(member.package_id())));
+
+        let missing_packages_with_the_features = unselected_members
+            .into_iter()
+            .filter(|member| {
+                unknown
+                    .iter()
+                    .any(|feature| member.summary().features().contains_key(&**feature))
+            })
+            .map(|m| m.name())
+            .collect_vec();
+
+        let these_features = if unknown.len() == 1 {
+            "this feature"
+        } else {
+            "these features"
+        };
+        let mut msg = if let [singular] = &selected_members[..] {
+            format!(
+                "the package '{}' does not contain {these_features}: {}",
+                singular.name(),
                 unknown.join(", ")
-            );
+            )
         } else {
-            bail!(
-                "none of the selected packages contains these features: {}, did you mean: {}?",
-                unknown.join(", "),
-                suggestions.join(", ")
+            let names = selected_members.iter().map(|m| m.name()).join(", ");
+            format!("none of the selected packages contains {these_features}: {}\nselected packages: {names}", unknown.join(", "))
+        };
+
+        use std::fmt::Write;
+        if !missing_packages_with_the_features.is_empty() {
+            write!(
+                &mut msg,
+                "\nhelp: include package{} with the missing feature{}:",
```

Comment:
I feel like this better matches the rustc style of "saying options exist"
```suggestion
                "\nhelp: package{} with the missing feature{}:",
```

Reply:
The suggestion includes the `-p` flag which I think is helpful, but without the verb it's weird to have flags there.

Reply:
You're right, I had missed the `-p`.

For myself, I'd prefer dropping `include` and dropping `--package`.

I feel like `include packages with the following features" is an imperative mood (we are telling the user what to do) which doesn't feel appropriate in this circumstance (or most).  

rustc's style guide does offer

> The message may contain further instruction such as "to do xyz, use" or "to do xyz, use abc".

In this case that would be "to include a package with the missing features, add one of these..."

I feel like that is best used when we are teaching someone which I don't think is needed here and it should build on information we've already mentioned which would be that packages exist with this feature but that comes later.

Reply:
As an example where telling the user what to do, I commonly have features through most of my workspace and I might leave it enabled when building on a package without it.  My action item would be to remove the feature, not to specify a different package.